### DISCUSSION TITLE
Which provider to use with Hardhat?

### DISCUSSION THREAD

**Discussion by ijonas** [wevm/wagmi#126] [Q&A] [answered]
Hi,

I'm trying to connect my wagmi-based app to a local Hardhat node, e.g. on http://localhost:8545.

I'd like to use Metamask as part of the development experience.

Should I use the InjectedConnector (with Metamask pointing to http://localhost:8545) or us the JsonRPCProvider pointing to http://localhost:8545?

Many thanks,
Ijonas.

**Comment by tmm** [comment, marked as answer]
`InjectedConnector` configured with `chain.hardhat` should work fine:

```ts
import { chain } from 'wagmi'
import { InjectedConnector } from 'wagmi/connectors/injected'

const connector = new InjectedConnector({
  chains: [chain.hardhat],
})
```

**Reply by ijonas** [reply]
Thank you so much for getting back to me so quickly.

**Comment by fmhall** [comment]
Hey @tmm, awesome work! Having some trouble with the above:

```typescript
import { chain } from 'wagmi'
import { InjectedConnector } from 'wagmi/connectors/injected'

const connector = new InjectedConnector({
  chains: [chain.hardhat],
})

<Provider
      connectors={[connector]}
    >
      <NextHead>
        <title>wagmi</title>
      </NextHead>

      <Component {...pageProps} />
</Provider>
```
It thinks it is connected to hardhat, but it is displaying data from ethereum mainnet. I have a feeling this is related to the `getDefaultProvider` call reverting back to eth mainnet.

How do we properly override the provider?

**Reply by tmm** [reply]
You should pass your own ethers.js Provider and use the `JsonRpcProvider` provider (pseudo-code below):

```tsx
import { providers } from 'ethers.js'
import { chain } from 'wagmi'
import { InjectedConnector } from 'wagmi/connectors/injected'

type ProviderConfig = {
  chainId?: number;
}

const provider = ({ chainId }: ProviderConfig) => {
  if (chainId === chain.hardhat.id)
    return new providers.JsonRpcProvider(chain.hardhat.rpcUrls[0])
  return providers.getDefaultProvider(chainId)
}

const connectors = [
  new InjectedConnector({
    chains: [chain.hardhat],
  }),
]

<Provider connectors={connectors} provider={provider>
  <Component {...pageProps} />
</Provider>
```
